api/v2/routes_rtfs.go

```diff
@@ -311,5 +354,73 @@ func (api *API) getDagObject(c *gin.Context) {
 		api.LogError(c, err, eh.IPFSDagGetError)(http.StatusBadRequest)
 		return
 	}
+	api.l.Infow("ipfs dag get requested", "user", username)
 	Respond(c, http.StatusOK, gin.H{"response": out})
 }
+
+// extendPin is used to extend the lifetime of a pin by a certain number of months
+func (api *API) extendPin(c *gin.Context) {
+	username, err := GetAuthenticatedUserFromContext(c)
+	if err != nil {
+		api.LogError(c, err, eh.NoAPITokenError)(http.StatusBadRequest)
+		return
+	}
+	// hash to retrieve dag for
+	hash := c.Param("hash")
+	if _, err := gocid.Decode(hash); err != nil {
+		Fail(c, err)
+		return
+	}
+	// extract needed post forms
+	forms := api.extractPostForms(c, "hold_time")
+	if len(forms) == 0 {
+		return
```

Comment:
don't we need a proper call to `Fail` or something of the sort for the request to end properly?

Reply:
Hmm that's a good point, I suppose this is a bit unclear actually since the "failure" is sent within `api.extractPostForms`

```golang
func (api *API) extractPostForms(c *gin.Context, formNames ...string) map[string]string {
	forms := make(map[string]string)
	for _, name := range formNames {
		value, exists := c.GetPostForm(name)
		if !exists {
			FailWithMissingField(c, name)
			return nil
		}
		forms[name] = value
	}
	return forms
}
```

let me refactor this a bit to be more clear

Reply:
yeah functions with hidden side effects like that are probably a bad idea - i would return an error with `extractPostForms` isntead

Reply:
oh your new implementation with `missingField` seems okay too 🆗